alf/trainers/evaluator.py

```diff
@@ -0,0 +1,259 @@
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from absl import logging
+from absl import flags
+import queue
+import torch.multiprocessing as mp
+import traceback
+import os
+import sys
+import time
+import torch
+from typing import Dict, Optional
+
+import alf
+from alf.algorithms.rl_algorithm import RLAlgorithm
+from alf.environments.alf_environment import AlfEnvironment
+from alf.utils import common
+from alf.data_structures import StepType
+from alf.algorithms.data_transformer import create_data_transformer
+from alf.environments.utils import create_environment
+from alf.trainers import policy_trainer
+from collections import namedtuple
+
+EvalJob = namedtuple(
+    "EvalJob", ["type", "global_counter", "step_metrics", "state_dict"],
+    defaults=[None] * 4)
+
+
+class AsyncEvaluator(object):
+    """Evaluate algorithm performance asynchronously in a different process.
+
+    Args:
+        conf_file: path of the config file
+        root_dir: directory for saving summary and checkpoints
+        num_evns: the number of environments for evaluating
+        seed: random seed
+    """
+
+    def __init__(self,
+                 conf_file: str,
+                 root_dir: str,
+                 num_envs: int,
+                 seed: Optional[int] = None):
+        ctx = mp.get_context('spawn')
+        self._job_queue = ctx.Queue()
+        self._done_queue = ctx.Queue()
+        self._worker = ctx.Process(
+            target=_worker,
+            args=(self._job_queue, self._done_queue, conf_file, num_envs,
+                  root_dir, seed))
+        self._worker.start()
+
+    def eval(self, algorithm: RLAlgorithm, step_metric_values: Dict[str, int]):
+        """Do one round of evaluation.
+
+        This function will return once the evaluator worker makes a copy of the
+        state_dict of ``algorithm``. However, if the previous evaluation has not
+        been finished, it will wait until it is finished.
+
+        The evaluation result will be written to log file and tensorboard by the
+        evaluation worker.
+
+        Args:
+            algorithm: the training algorithm
+            step_metric_values: a dictionary of step metric values to generate
+                the evaluation summaries against. Note that it needs to contain
+                "EnvironmentSteps" at least.
+        """
+        job = EvalJob(
+            type="eval",
+            step_metrics=step_metric_values,
+            global_counter=int(alf.summary.get_global_counter()),
+            state_dict=algorithm.state_dict())
+        self._job_queue.put(job)
+        self._done_queue.get()
```

Comment:
When a previous eval job is still running and eval is called again, the worker is still busy working on the previous job and cannot yet take the job off the queue, so in this case, the evaluator will block on this done_queue.get()?

Maybe let's add a timer here to log to tensorboard how long this done_queue.get() is taking, so that we'll know that the eval process is blocking the training.

Reply:
added